[MRG] Pass sample_weight when predicting on stacked folds

wderose opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #16537.

#### What does this implement/fix? Explain your changes.
This PR passes `sample_weight` via `fit_params` when we call `cross_val_predict` in `_BaseStacking`, as per the comment [here](https://github.com/scikit-learn/scikit-learn/issues/16537#issuecomment-590573769).
https://github.com/scikit-learn/scikit-learn/blob/136ef7988402aee1bd839eba21ea989e2c1e77cb/sklearn/ensemble/_stacking.py#L169-L175

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->
cc: @caioaao @glemaitre

Commit:
pass sample_weight when predicting on stacked folds

Commit:
use more descriptive spec def

jnothman commented on sklearn/ensemble/tests/test_stacking.py:
```diff
@@ -439,6 +440,19 @@ def test_stacking_with_sample_weight(stacker, X, y):
 
 
+def test_stacking_classifier_sample_weight_fit_param():
+    # check sample_weight is passed to all invokations of fit
```
```suggestion
    # check sample_weight is passed to all invocations of fit
```

jnothman reviewed (approved):
Please add a `|Fix|` entry to the change log at `doc/whats_new/v0.23.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`

Please add a versionchanged note in the docstring of sample_weight

jnothman commented:
But tests are failing. Should have checked that first

Commit:
fix specs by conditionally passing sample_weight

Commit:
Add entry in release notes describing improvement

Commit:
add versionchanged note about sample_weight

Commit:
fixup witespace in docstring

Commit:
indent bullet list in release notes

Commit:
Merge remote-tracking branch 'upstream/master' into is/16537

wderose commented:
> But tests are failing. Should have checked that first

@jnothman : Thank you for the prompt review.

I've added the requested doc changes and fixed the failing specs.

Please let me know if you'd like more details in the versionchange/release notes or if you'd like to see more tests.

glemaitre commented on doc/whats_new/v0.23.rst:
```diff
@@ -179,6 +179,12 @@ Changelog
   used during `fit`.
   :pr:`16437` by :user:`Jin-Hwan CHO <chofchof>`.
 
+- |Fix| Fixed a bug in :class:`ensemble._BaseStacking` where the
```
Instead of pointing to BaseStacking point to StackingClassifier and StackingRegressor since they will be the class used by our users

glemaitre commented on sklearn/ensemble/_stacking.py:
```diff
@@ -165,10 +169,13 @@ def fit(self, X, y, sample_weight=None):
-
+        fit_params = {"sample_weight": sample_weight} \
+            if sample_weight is not None \
+            else None
```
```suggestion
        fit_params = ({"sample_weight": sample_weight}
                      if sample_weight is not None
                      else None)
```

glemaitre commented on sklearn/ensemble/_stacking.py:
```diff
@@ -182,21 +189,8 @@ def fit(self, X, y, sample_weight=None):
-                raise
-        else:
-            self.final_estimator_.fit(X_meta, y)
+        _parallel_fit_estimator(self.final_estimator_, X_meta, y,
```
I would put a comment because `_parallel` is a weird name since we fit a single estimator but this is worth.
However, I would propose a subsequent maintenance PR to rename `_parallel_fit_estimator` to `_fit_single_estimator`.
The parallel part will come when we call `Parallel` on the `fit_single_estimator`

wderose replied:
Good idea, I thought this naming was a little strange as well.

I added a TODO comment to the docstring of `_parallel_fit_estimator` and linked to this review comment. It seems like a simple enough change that I may tackle it as a maintenance fix in a follow-up PR.

wderose replied:
Created a follow-up issue for reference in the maintenance PR that will do the renaming: https://github.com/scikit-learn/scikit-learn/issues/16595

wderose replied:
Change made in this PR: https://github.com/scikit-learn/scikit-learn/pull/16599

glemaitre reviewed (commented):
This looks good. Only a couple of comments

Commit:
fix: update release notes to reference to StackingClassifier and StackingRegressor instead of _BaseStacking

Commit:
fix: minor formatting suggestion from review comments

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

Commit:
fix: add comment describing odd naming of _parallel_fit_estimator

Commit:
fix: update link to PR with direct link to issue

wderose commented:
@glemaitre Thank you for the comments. 

I updated the release notes to reference `StackingClassifier` and `StackingRegressor` directly. 

I also created a follow-on issue (#16595) regarding the naming of `_parallel_fit_estimator` and added a comment noting the strange naming.

Please take another look at these changes and let me know if there is anything else you'd like me to include.

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

Commit:
set n_features_in_ when CheckingClassifier is fit

wderose commented on sklearn/utils/_mocking.py:
```diff
@@ -96,6 +96,7 @@ def fit(self, X, y, **fit_params):
             assert self.check_X(X)
         if self.check_y is not None:
             assert self.check_y(y)
+        self._check_n_features(X, reset=True)
```
I needed to add this call so that `n_features_in_` is set correctly in `CheckingClassifier`. 

The property is read during the fit method of BaseStacking: https://github.com/scikit-learn/scikit-learn/commit/d205638475ca542dc46862652e3bb0be663a8eac#diff-c8f397939f2393da5b2eef4285761da1R143

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

Commit:
fix formatting

Commit:
use len(X) as n_features_in_

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

Commit:
Merge branch 'master' of https://github.com/scikit-learn/scikit-learn into is/16537

thomasjpfan reviewed (approved):
Thank you for the PR @wderose 

LGTM!
